source/adios2/toolkit/burstbuffer/FileDrainer.cpp

```diff
@@ -0,0 +1,279 @@
+    std::lock_guard<std::mutex> lockGuard(operationsMutex);
+    operations.push(operation);
+}
+
+void FileDrainer::AddOperationSeekEnd(const std::string &toFileName)
+{
+    std::string emptyStr;
+    AddOperation(DrainOperation::SeekEnd, emptyStr, toFileName, 0, 0, 0);
+}
+void FileDrainer::AddOperationCopyAt(const std::string &fromFileName,
+                                     const std::string &toFileName,
+                                     size_t fromOffset, size_t toOffset,
+                                     size_t countBytes)
+{
+    AddOperation(DrainOperation::CopyAt, fromFileName, toFileName, fromOffset,
+                 toOffset, countBytes);
+}
+void FileDrainer::AddOperationCopy(const std::string &fromFileName,
+                                   const std::string &toFileName,
+                                   size_t countBytes)
+{
+    AddOperation(DrainOperation::Copy, fromFileName, toFileName, 0, 0,
+                 countBytes);
+}
+
+void FileDrainer::AddOperationWriteAt(const std::string &toFileName,
+                                      size_t toOffset, size_t countBytes,
+                                      const void *data)
+{
+    std::string emptyStr;
+    AddOperation(DrainOperation::WriteAt, emptyStr, toFileName, 0, toOffset,
+                 countBytes, data);
+}
+
+void FileDrainer::AddOperationWrite(const std::string &toFileName,
+                                    size_t countBytes, const void *data)
+{
+    std::string emptyStr;
+    AddOperation(DrainOperation::Write, emptyStr, toFileName, 0, 0, countBytes,
+                 data);
+}
+
+void FileDrainer::AddOperationOpen(const std::string &toFileName, Mode mode)
+{
+    std::string emptyStr;
+    if (mode == Mode::Write)
+    {
+        AddOperation(DrainOperation::Create, emptyStr, toFileName, 0, 0, 0);
+    }
+    else if (mode == Mode::Append)
+    {
+        AddOperation(DrainOperation::Open, emptyStr, toFileName, 0, 0, 0);
+    }
+    else
+    {
+        throw std::runtime_error(
+            "ADIOS Coding ERROR: FileDrainer::AddOperationOpen() only supports "
+            "Write and Append modes\n");
+    }
+}
+
+InputFile FileDrainer::GetFileForRead(const std::string &path)
+{
+    auto it = m_InputFileMap.find(path);
+    if (it != m_InputFileMap.end())
+    {
+        return it->second;
+    }
+    else
+    {
+        InputFile f = std::make_shared<std::ifstream>();
+        m_InputFileMap.emplace(path, f);
+        Open(f, path);
+        return f;
+    }
+}
+
+OutputFile FileDrainer::GetFileForWrite(const std::string &path, bool append)
+{
+    auto it = m_OutputFileMap.find(path);
+    if (it != m_OutputFileMap.end())
+    {
+        return it->second;
+    }
+    else
+    {
+        OutputFile f = std::make_shared<std::ofstream>();
+        m_OutputFileMap.emplace(path, f);
+        Open(f, path, append);
+        return f;
+    }
+}
+
+void FileDrainer::Open(InputFile f, const std::string &path)
```

Comment:
Since `InputFile` is a `shared_ptr` this add to the ownership counter. Ownership is not needed so passing `InputFile &`  should be fine.